DuckDuckGo/TabViewController.swift

```diff
@@ -1650,12 +1546,324 @@ extension TabViewController: WKNavigationDelegate {
+            Pixel.fire(pixel: .downloadAttemptToOpenBLOBviaJS)
+            self.legacySetupBlobDownload(for: navigationAction) {
+                completion(.allow)
+            }
+            return
+        }
+
+        self.blobDownloadTargetFrame = navigationAction.targetFrame
+        completion(.allow)
+    }
+
+    @discardableResult
+    private func startDownload(with navigationResponse: WKNavigationResponse,
+                               decisionHandler: @escaping (WKNavigationResponsePolicy) -> Void) -> Download? {
+        let downloadManager = AppDependencyProvider.shared.downloadManager
+        let cookieStore = webView.configuration.websiteDataStore.httpCookieStore
+        let url = navigationResponse.response.url!
+
+        if case .blob = SchemeHandler.schemeType(for: url) {
+            if #available(iOS 14.5, *) {
+                decisionHandler(.download)
+
+                return nil
+
+            // [iOS<14.5 legacy] reuse temporary download for blob: initiated by WKNavigationAction
+            } else if let download = self.temporaryDownloadForPreviewedFile,
+                      download.temporary,
+                      download.url == navigationResponse.response.url {
+                self.temporaryDownloadForPreviewedFile = nil
+                download.temporary = FilePreviewHelper.canAutoPreviewMIMEType(download.mimeType)
+                downloadManager.startDownload(download)
+
+                decisionHandler(.cancel)
+
+                return download
+            }
+        } else if let download = downloadManager.makeDownload(navigationResponse: navigationResponse, cookieStore: cookieStore) {
+            downloadManager.startDownload(download)
+            decisionHandler(.cancel)
+
+            return download
+        }
+
+        decisionHandler(.cancel)
+        return nil
+    }
+
+    /*
+     Some files might be previewed by webkit but in order to share them
+     we need to download them first.
+     This method stores the temporary download or clears it if necessary
+     */
+    private func setupOrClearTemporaryDownload(for response: URLResponse,
+                                               completion: @escaping (WKNavigationResponsePolicy) -> Void) {
+        let downloadManager = AppDependencyProvider.shared.downloadManager
+        guard let url = response.url,
+              let downloadMetaData = downloadManager.downloadMetaData(for: response),
+              !downloadMetaData.mimeType.isHTML
+        else {
+            temporaryDownloadForPreviewedFile?.cancel()
+            temporaryDownloadForPreviewedFile = nil
+            completion(.allow)
+            return
+        }
+        guard SchemeHandler.schemeType(for: url) != .blob else {
+            // suggestedFilename is empty for blob: downloads unless handled via completion(.download)
+            // WKNavigationResponse._downloadAttribute private API could be used instead of it :(
```

Comment:
@mallexxx @miasma13 do you think we should file a ~~radar~~ feedback with Apple?

Reply:
I don‘t think so, I'd rather switch to WKDownload usage instead of custom downloads implementation when we drop iOS <14.5